src/EditorFeatures/Core/Implementation/IntelliSense/EditorCompletionImplementation/EditorAsyncCompletionItemManager.cs

```diff
@@ -0,0 +1,670 @@
+        private ImmutableArray<string> _recentItems = ImmutableArray<string>.Empty;
+
+        public EditorAsyncCompletionItemManager(IAsyncCompletionBroker broker)
+        {
+            _broker = broker;
+            _completionHelper = new CompletionHelper(isCaseSensitive: true);
+        }
+
+        // This method is supposed to be called just once per session. That is why we register events there.
+        // If the idea is changed, we need an Initialize method to be added to the interface.
+        public Task<ImmutableArray<EditorCompletion.CompletionItem>> SortCompletionListAsync(
+            IAsyncCompletionSession session,
+            EditorCompletion.AsyncCompletionSessionInitialDataSnapshot data,
+            CancellationToken token)
+        {
+            session.ItemCommitted += ItemCommitted;
+            session.Dismissed += SessionDismissed;
+            return Task.FromResult(data.InitialList.OrderBy(i => i.SortText).ToImmutableArray());
+        }
+
+        private void ItemCommitted(object sender, EditorCompletion.CompletionItemEventArgs e)
+        {
+            MakeMostRecentItem(e.Item.DisplayText);
+        }
+
+        private void SessionDismissed(object sender, EventArgs e)
+        {
+            if (sender is IAsyncCompletionSession session)
+            {
+                session.ItemCommitted -= ItemCommitted;
+                session.Dismissed -= SessionDismissed;
+            }
+        }
+
+        public Task<EditorCompletion.FilteredCompletionModel> UpdateCompletionListAsync(
+            IAsyncCompletionSession session,
+            EditorCompletion.AsyncCompletionSessionDataSnapshot data,
+            CancellationToken cancellationToken)
+        {
+            var mustSetSelection = false;
+
+            var filterText = session.ApplicableToSpan.GetText(data.Snapshot);
+
+            // Check if the user is typing a number. If so, only proceed if it's a number
+            // directly after a <dot>. That's because it is actually reasonable for completion
+            // to be brought up after a <dot> and for the user to want to filter completion
+            // items based on a number that exists in the name of the item. However, when
+            // we are not after a dot (i.e. we're being brought up after <space> is typed)
+            // then we don't want to filter things. Consider the user writing:
+            //
+            //      dim i =<space>
+            //
+            // We'll bring up the completion list here (as VB has completion on <space>).
+            // If the user then types '3', we don't want to match against Int32.
+            if (filterText.Length > 0 && char.IsNumber(filterText[0]))
+            {
+                if (!IsAfterDot(data.Snapshot, session.ApplicableToSpan))
+                {
+                    return Task.FromResult(new EditorCompletion.FilteredCompletionModel(ImmutableArray<EditorCompletion.CompletionItemWithHighlight>.Empty, 0));
```

Comment:
Should the editor provide a FilteredCompletionModel.Empty?

Reply:
@amwieczo@microsoft.com please take a look

---
In reply to: [208708360](https://github.com/dotnet/roslyn/pull/29016#discussion_r208708360) [](ancestors = 208708360)